lib/IntermediateRepresentation/afforestation.cc

```diff
@@ -0,0 +1,1963 @@
+        common::visitors{
+            [](const parser::LoopBounds<parser::ScalarIntExpr> &loopBounds) {
+              auto *loopVariable{loopBounds.name.thing.thing.symbol};
+              auto *second{loopBounds.upper.thing.thing->typedExpr.get()};
+              if (loopBounds.step.has_value()) {
+                auto *step{loopBounds.step->thing.thing->typedExpr.get()};
+                SEMANTICS_CHECK(step, "DO step expression missing");
+                if (evaluate::IsConstantExpr(step->v)) {
+                  auto stepValue{evaluate::ToInt64(step->v)};
+                  SEMANTICS_CHECK(stepValue != 0, "step cannot be zero");
+                  if (stepValue > 0) {
+                    return LessThanOrEqualToComparison(loopVariable, second);
+                  }
+                  return GreaterThanOrEqualToComparison(loopVariable, second);
+                }
+                // expr ::= (step > 0 && v <= snd) || (step < 0 && v >= snd)
+                return AlwaysTrueExpression();  // FIXME
+              }
+              return LessThanOrEqualToComparison(loopVariable, second);
+            },
+            [](const parser::ScalarLogicalExpr &scalarLogicalExpr) {
+              auto &expression{scalarLogicalExpr.thing.thing};
+              SEMANTICS_CHECK(
+                  expression->typedExpr.get(), "DO WHILE condition missing");
+              return expression->typedExpr.get();
+            },
+            [](const parser::LoopControl::Concurrent &concurrent) {
+              return AlwaysTrueExpression();  // FIXME
+            },
+        },
+        loopControl->u);
+  }
+  return AlwaysTrueExpression();
+}
+
+static void CreateSwitchHelper(IntermediateRepresentationBuilder *builder,
+    const Evaluation &condition, BasicBlock *defaultCase,
+    const SwitchStmt::ValueSuccPairListType &rest) {
+  builder->CreateSwitch(condition, defaultCase, rest);
+}
+static void CreateSwitchCaseHelper(IntermediateRepresentationBuilder *builder,
+    const Evaluation &condition, BasicBlock *defaultCase,
+    const SwitchCaseStmt::ValueSuccPairListType &rest) {
+  builder->CreateSwitchCase(condition, defaultCase, rest);
+}
+static void CreateSwitchRankHelper(IntermediateRepresentationBuilder *builder,
+    const Evaluation &condition, BasicBlock *defaultCase,
+    const SwitchRankStmt::ValueSuccPairListType &rest) {
+  builder->CreateSwitchRank(condition, defaultCase, rest);
+}
+static void CreateSwitchTypeHelper(IntermediateRepresentationBuilder *builder,
+    const Evaluation &condition, BasicBlock *defaultCase,
+    const SwitchTypeStmt::ValueSuccPairListType &rest) {
+  builder->CreateSwitchType(condition, defaultCase, rest);
+}
+
+struct FortranIRLowering {
+  using LabelMapType = std::map<LinearLabelRef, BasicBlock *>;
+  using Closure = std::function<void(const LabelMapType &)>;
+
+  FortranIRLowering(semantics::SemanticsContext &sc, bool debugLinearIR)
+    : fir_{new Program("program_name")}, semanticsContext_{sc},
```

Comment:
Is `program_name` a placeholder for something that should be passed in as an argument or a constituent of the SemanticsContext?

Reply:
Yes, it is just a placeholder for now. It could be taken from a PROGRAM statement (or like) if available.